aQute.libg/src/aQute/lib/concurrent/serial/SerialExecutor.java

```diff
@@ -0,0 +1,90 @@
+package aQute.lib.concurrent.serial;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+
+import org.osgi.util.promise.Deferred;
+import org.osgi.util.promise.Promise;
+
+/**
+ * Will execute a set of tasks in order of submit.
+ */
+public class SerialExecutor implements Closeable {
+	final Executor			executor;
+	final List<Runnable>	tasks	= new ArrayList<>();
+	volatile Thread			thread;
+
+	/**
+	 * The executor to use.
+	 *
+	 * @param executor
+	 */
+	public SerialExecutor(Executor executor) {
+		this.executor = executor;
+	}
+
+	/**
+	 * Run serial in order of submission and return a promise
+	 *
+	 * @param <T> the return type
+	 * @param callable the callable providing the data
+	 * @return the promise
+	 */
+	public <T> Promise<T> submit(Callable<T> callable) {
+		Deferred<T> deferred = new Deferred<>();
```

Comment:
Since you are supplied an executor in the construtor, you should probably create a PromiseFactory using that executor and use the promise factory to create the deferred. Otherwise callbacks on the returned promise can occur on the default executor for promises.

Reply:
ok